source/h2_frames.c

```diff
@@ -332,714 +269,900 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
+};
 
-    AWS_ZERO_STRUCT(*frame);
-    frame->base.type = AWS_H2_FRAME_T_PING;
+DEFINE_FRAME_VTABLE(prebuilt);
 
-    return AWS_OP_SUCCESS;
-}
-void aws_h2_frame_ping_clean_up(struct aws_h2_frame_ping *frame) {
-    AWS_PRECONDITION(frame);
-
-    (void)frame;
+/* Can't pre-encode a frame unless it's guaranteed to fit, regardless of current settings. */
+static size_t s_prebuilt_payload_max(void) {
+    return aws_h2_settings_bounds[AWS_H2_SETTINGS_MAX_FRAME_SIZE][0];
 }
 
-int aws_h2_frame_ping_encode(
-    struct aws_h2_frame_ping *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
+/* Create aws_h2_frame_prebuilt and encode frame prefix into frame->encoded_buf.
+ * Caller must encode the payload to fill the rest of the encoded_buf. */
+static struct aws_h2_frame_prebuilt *s_h2_frame_new_prebuilt(
+    struct aws_allocator *allocator,
+    enum aws_h2_frame_type type,
+    uint32_t stream_id,
+    size_t payload_len,
+    uint8_t flags) {
+
+    if (payload_len > s_prebuilt_payload_max()) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
+    }
 
-    (void)encoder;
+    const size_t encoded_frame_len = s_frame_prefix_length + payload_len;
 
-    if (frame->base.stream_id != 0) {
-        return aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
+    /* Use single allocation for frame and buffer storage */
+    struct aws_h2_frame_prebuilt *frame;
+    void *storage;
+    if (!aws_mem_acquire_many(
+            allocator, 2, &frame, sizeof(struct aws_h2_frame_prebuilt), &storage, encoded_frame_len)) {
+        return NULL;
     }
 
-    const size_t output_init_len = output->len;
+    AWS_ZERO_STRUCT(*frame);
+    s_init_frame_base(&frame->base, allocator, type, &s_frame_prebuilt_vtable, stream_id);
+    frame->encoded_buf = aws_byte_buf_from_empty_array(storage, encoded_frame_len);
+    frame->send_progress = aws_byte_cursor_from_array(storage, encoded_frame_len);
 
-    /* Write the header data */
-    uint8_t flags = 0;
-    if (frame->ack) {
-        flags |= AWS_H2_FRAME_F_ACK;
+    if (s_frame_prefix_encode(type, stream_id, payload_len, flags, &frame->encoded_buf)) {
+        goto error;
     }
 
-    if (s_frame_prefix_encode(&frame->base, AWS_H2_PING_DATA_SIZE, flags, output)) {
-        goto write_error;
+    return frame;
+
+error:
+    s_frame_prebuilt_destroy(&frame->base);
+    return NULL;
+}
+
+static void s_frame_prebuilt_destroy(struct aws_h2_frame *frame_base) {
+    aws_mem_release(frame_base->alloc, frame_base);
+}
+
+static int s_frame_prebuilt_encode(
+    struct aws_h2_frame *frame_base,
+    struct aws_h2_frame_encoder *encoder,
+    struct aws_byte_buf *output,
+    bool *complete) {
+
+    struct aws_h2_frame_prebuilt *frame = AWS_CONTAINER_OF(frame_base, struct aws_h2_frame_prebuilt, base);
+
+    if (frame->send_progress.len == frame->encoded_buf.len) {
```

Comment:
I'm a little lost, intuitively I thought "wouldn't these both be 0?", but then down below this branch send_progress.len is non-zero, so I don't understand, maybe a comment would help?

I'm sure when it's explained I'm gonna go "OH RIGHT"